runtime/src/accounts_background_service.rs

```diff
@@ -169,23 +174,60 @@ impl SnapshotRequestHandler {
                 }
 
+                let block_height = snapshot_root_bank.block_height();
+                let snapshot_type = if block_height
+                    % self.snapshot_config.full_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::FullSnapshot)
+                } else if block_height
+                    % self
+                        .snapshot_config
+                        .incremental_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::IncrementalSnapshot(last_full_snapshot_slot))
+                } else {
+                    None
+                };
+
                 // Generate an accounts package
+                let mut attempts = 0;
                 let mut snapshot_time = Measure::start("snapshot_time");
-                let r = snapshot_utils::snapshot_bank(
-                    &snapshot_root_bank,
-                    status_cache_slot_deltas,
-                    &self.accounts_package_sender,
-                    &self.snapshot_config.bank_snapshots_dir,
-                    &self.snapshot_config.snapshot_archives_dir,
-                    self.snapshot_config.snapshot_version,
-                    &self.snapshot_config.archive_format,
-                    hash_for_testing,
-                );
-                if r.is_err() {
+                loop {
+                    attempts += 1;
+                    let result = snapshot_utils::snapshot_bank(
+                        &snapshot_root_bank,
+                        status_cache_slot_deltas.clone(),
+                        &self.accounts_package_sender,
+                        &self.snapshot_config.bank_snapshots_dir,
+                        &self.snapshot_config.snapshot_archives_dir,
+                        self.snapshot_config.snapshot_version,
+                        self.snapshot_config.archive_format,
+                        hash_for_testing,
+                        snapshot_type,
+                    );
+
+                    if result.is_ok() {
+                        break;
+                    }
+
                     warn!(
-                        "Error generating snapshot for bank: {}, err: {:?}",
+                        "Error generating snapshot for bank. attempts: {}, slot: {}, err: {:?}",
+                        attempts,
                         snapshot_root_bank.slot(),
-                        r
+                        result
+                    );
+
+                    // Only require success for full snapshots.  Otherwise just log the error and
+                    // continue on.
+                    if snapshot_type != Some(SnapshotType::FullSnapshot) {
+                        break;
+                    }
+
+                    assert!(
+                        attempts <= SNAPSHOT_BANK_MAX_ATTEMPTS,
+                        "Unable to snapshot bank intended for full snapshot!"
```

Comment:
Do we want to panic? Earlier we only logged an error

Reply:
Correct! There is new behavior; I put it in the description, but maybe I need it in the code too?

>!!! New behavior changes !!!
>
>Full snapshots (both bank and archive) **MUST** succeed.  If there are
>failures, the functions will be retried.  If they still fail, an assert
>will be triggered.
>
>This is required because of how the last full snapshot slot is
>calculated, which is used by AccountsBackgroundService when calling
>`clean_accounts()`.

Reply:
Heh, and now this has changed again. There's no more retrying in ABS, so this is all gone now.